### PR TITLE
feat(cake_nvfp4_attn): add experimental NVFP4 attention on SM103

### PR DESCRIPTION
**Pull request by yyihuang**
## 📌 Description

Add experimental NVFP4 attention on SM103 through
`flashinfer.prefill.prepare_nvfp4_attention(q, k, v, out, backend="cake")`.
Preparation quantizes contiguous BF16 Q/K/V once and returns an
`NVFP4AttentionRunner`. Calling the runner executes QK, softmax and PV attention
and writes into the caller's BF16 output without CUDA allocation. Prepare a
new runner when input values or bindings change. The initial backend supports
noncausal attention with D=128 and S divisible by 512.

The API emits the standard experimental warning and hands off to the feature
package. Preparation bounds E4M3 block scales to their finite range before
conversion, including on PyTorch versions that do not saturate finite overflow.
Preparation, JIT registration, and generated CUDA are contained under
`flashinfer/experimental/nvfp4_attention/`.

On all three validated shapes, export latency is within 3% of both source
launch policies. The largest measured increase is **0.2228%** versus
selected source and **0.3008%** versus the original pointer-argument
policy, both rebuilt with the current compiler. These are prepared-submit
measurements; input quantization and preparation are outside timing. The
generated CUDA and exported launch code are unchanged.

## Generated-program export evidence

**Updated on 2026-09-19: NVIDIA GB300 (SM103), 2386.7–3099.3 achieved TFLOP/s.**
The complete three-shape campaign now passes in **one process**, with the real
PyTorch CPU/CUDA allocation profiler running **before each row's timing**.
Correctness, zero allocation/free, strict CUPTI cold-L2 timing, and the
unchanged per-shape 3% performance gates all pass.

The original absolute readings were invalid because allocation profiling
replaced process-global CUPTI activity-buffer and timestamp callbacks.
Restoring only activity-buffer callbacks recovered records but could leave
CPU clock ticks interpreted as nanoseconds. The benchmark collector used for
this validation re-registers its activity callbacks and restores a native
nanosecond timestamp callback for each collection session. Its native lifetime
also permits clean CUDA context teardown. It does not rescale old results or
fall back to another GPU timer.

This removes the earlier fresh-process/profiler-after-timing workaround.
The exported kernel was not changed to obtain these numbers. The selected
source uses grid-constant TMA arguments and PDL; the original policy uses
pointer arguments without PDL. Both policies are rebuilt from unchanged
kernel source using the current compiler. The historical compiler snapshot
was unavailable, so this is not an exact frozen-compiler rerun.

## Hardware and protocol

- One exclusively allocated NVIDIA GB300, SM103 / capability 10.3;
  driver 580.159.03, PyTorch 2.13.0a0+9186a08b2c.nv26.07, CUDA 13.3.
- Unchanged target revision: `034bb09742566ff991abad1a56dd28f5e8efac2f`.
- CUPTI activity timing, cold L2, fallback disabled; symmetric external
  CUDA Graphs, three counterbalanced groups, 100 ms warmup and 1000 ms
  reportable budget per arm/group. SM clock threshold: 1900 MHz.
- Selected source, original source policy, and export share the same physical
  input/output buffers within each row. All three pass independent GPU SDPA
  checks over quantized Q/K/V before and after timing at `atol=1, rtol=0.1`.
  Independently prepared packed inputs match byte-for-byte and remain unchanged.
- Every warmed export submit records zero device allocations/frees and zero
  allocation/free API events, with the expected native launch and kernel present.

## Per-shape results

| Shape (B,H,S,D) | Selected source ms | Export ms | Source / Export | Original source policy ms | Original / Export | Correctness and gates |
|---|---:|---:|---:|---:|---:|---|
| (4,8,4096,128) | 0.114913 | **0.115169** | 0.997777x | 0.115553 | 1.003334x | pass |
| (1,8,32768,128) | 1.437610 | **1.440234** | 0.998178x | 1.439690 | 0.999622x | pass |
| (8,32,8192,128) | 2.833268 | **2.838066** | 0.998309x | 2.829555 | 0.997001x | pass |

## Absolute attention throughput

Using the dense-attention FLOPs convention, QK and PV count two FLOPs per
multiply-add:

`FLOPs per call = 4 * B * H * S^2 * D`

`TFLOP/s = FLOPs per call / (latency_ms * 10^9)`

The numerator excludes softmax, scaling, and quantization operations. The
measured attention latency includes softmax and internal kernel work; input
Q/K/V quantization and preparation are outside timing. These are achieved
useful TFLOP/s with no sparse-operation multiplier.

| Shape (B,H,S,D) | FLOPs per call | Selected source TFLOP/s | Export TFLOP/s | Original source policy TFLOP/s |
|---|---:|---:|---:|---:|
| (4,8,4096,128) | 274,877,906,944 | 2392.1 | **2386.7** | 2378.8 |
| (1,8,32768,128) | 4,398,046,511,104 | 3059.3 | **3053.7** | 3054.9 |
| (8,32,8192,128) | 8,796,093,022,208 | 3104.6 | **3099.3** | 3108.6 |

Geomeans over all three correct, timed rows: selected source **0.776425 ms**,
export **0.777912 ms**, source/export **0.998088x**;
original source policy **0.777898 ms**, original/export
**0.999983x**. All shapes carry the aligned,
noncausal, and owner-required tags, so those tag denominators are also 3/3.
No shape or gate-failing row was dropped. Correctness and both 3% performance
comparisons pass on **3/3 shapes**, along with the 2% direction/drift gates.

## Timing and isolation evidence

All three rows use the same worker PID, with allocation profiling before
each row. All **36 CUPTI collections** pass positive/nonoverlapping interval,
independent physical-envelope, and synchronized-completion checks. The
benchmark process exits normally, including interpreter teardown. Host
durations are consistency diagnostics, not substitutes for CUPTI GPU timings.

The approximately 1 Hz GPU/process census contains **276 samples**, with no
query errors and no observed process other than the expected worker on the
measured GPU. Samples at GPU utilization of at least 90% report **2070 MHz**.
These are sampled NVML clocks/processes, not continuous CUPTI clock measurements.
Before/after GPU reports, allocation profiles, and full CUPTI activity intervals
are retained with the validation results.

## Validation turnaround

The successful three-row validation took **297.075 s** managed end to end,
including **285.163 s** in the runner. Including source/environment setup and
preflight repairs, turnaround from the first setup attempt to successful
validation was **1836.306 s**. Completed native builds were reused. GPU kernel
latencies are reported separately above.

## 🔍 Related Issues

Tracking issue: [#4254](https://github.com/flashinfer-ai/flashinfer/issues/4254).
Owner: @yyihuang. The tracker records the use case, experimental rationale,
and proposed graduation target of v0.7.0, subject to maintainer review.

## 🚀 Pull Request Checklist

Thank you for contributing to FlashInfer! Before we review your pull request, please make sure the following items are complete.

### ✅ Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` (or used your preferred method).
- [x] I have installed the hooks with `pre-commit install`.
- [x] I have run the hooks manually with `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure about how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

## 🧪 Tests

- [x] Tests have been added or updated as needed.
- [x] All tests are passing (`unittest`, etc.).

The checked test result applies to the declared experimental test scope:

Latest naming and scale-range repair (`034bb09742566ff991abad1a56dd28f5e8efac2f`):

- `NVFP4AttentionRunner` names the execution object; calling it runs QK, softmax
  and PV attention on its bound inputs.
- `python -m pytest tests/experimental/test_cake_nvfp4_attention.py -q`:
  **6/6 passed** on B300 (SM103), including the original three shapes and
  CPU/CUDA checks for zero, normal values, the E4M3 scale limit, and signed
  BF16 finite extremes. The example and targeted Ruff checks also passed.
- Validation used PyTorch `2.13.0a0+9186a08b2c.nv26.07` / CUDA 13.3. This build
  already saturates the old quantizer on CPU and CUDA; the explicit scale bound
  also covers supported builds whose FP8 conversion can produce NaN on finite
  overflow.
- Test and example processes took **34.731 s** together. The successful managed
  validation took **95.666 s** end to end. Including two setup failures and
  their repairs, the validation campaign took **300.956 s** from first submission
  to successful completion. Those setup failures occurred before testing.

The corrected performance tables above measure this revision. The naming and
scale repair changes preparation-side behavior, with no generated CUDA or
launch-path changes.

Earlier export and packaging validation:

- `python -m pytest tests/experimental/test_cake_nvfp4_attention.py -q`: **4/4 passed**, covering the three shapes, output reuse, reference correctness, and rejection of an unknown backend.
- `python examples/cake_nvfp4_attention.py`: **passed**.
- The test and example processes import the target implementation without the source implementation on their Python import path.
- Canonical export validation: **3/3 passed**, including native build and package-data discovery, zero-allocation/free submit checks, and both per-row performance gates.

The successful tests/example workload took **16.490 s**; its managed turnaround
was **36.337 s**. This duration is separate from the export campaign above.

- [x] Formatting checks completed.
- [x] Wheel packaging and archive-content verification completed.

Wheel inspection verified eight members, including both generated CUDA translation
units and required headers, with source bytes matching the checkout. This checks
packaging; GPU validation used the source checkout. Wheel validation took
**222.754 s**, including **222.630 s** of packaging. The combined pre-commit and
wheel workload took **284.323 s**, with **287.694 s** managed turnaround.

## 🔬 Experimental Track

- [x] This PR is **experimental**: it adds or changes code under `flashinfer/experimental/` and/or an `@flashinfer_experimental_api`. Tracking issue: #4254
  - [x] The tracking issue names an owner, the reason for the experimental path, and a graduation plan with a target release.
  - [x] Core changes are limited to a thin entry point (signature, shared validation, feature-gate check, backend selection, handoff).
  - [x] Tests live in `tests/experimental/` and were validated on the intended hardware; a runnable example is included.
  - [x] Nothing is registered in `flashinfer/aot.py`, and no experimental backend is reachable from `backend="auto"` without `FLASHINFER_ALLOW_EXPERIMENTAL_AUTO_BACKENDS=1`. (Calling an `@flashinfer_experimental_api` or naming a backend explicitly is itself the opt-in and needs no environment variable.)
  - [x] **Test scope declared below.** The experimental CI lane runs exactly these targets, so keep them as narrow as the change allows.

```experimental-tests
tests/experimental/test_cake_nvfp4_attention.py
```

## Reviewer Notes

Both the API and backend are experimental. Review the preparation/submit
ownership contract, SM103 routing, and JIT source packaging. The validated
performance set is exactly the three noncausal shapes above. Formatting and
pre-commit and wheel packaging checks passed.

## Summary by CodeRabbit

* **New Features**
  * Added experimental NVFP4 attention support through the Cake backend for compatible GPUs.
  * Supports reusable prepared operations for contiguous BF16 inputs, with results written to caller-provided output tensors.
  * Added an example demonstrating NVFP4 attention usage and result validation.

* **Documentation**
  * Added guidance on supported inputs, limitations, backend behavior, and experimental API usage.

* **Bug Fixes**
  * Improved handling of large quantization scales by saturating finite overflow values.

### CODE REVIEW

**Comment by yyihuang** [comment]
@flashinfer-bot run tests/experimental/test_cake_nvfp4_attention.py

**Comment by yzh119** [comment]
@yyihuang can you also show the absolute flops number?

**Comment by yyihuang** [comment]
@flashinfer-bot run tests/experimental/test_cake_nvfp4_attention.py

**Comment by yyihuang** [comment]
@flashinfer-bot run tests/experimental/test_cake_nvfp4_attention.py

**Review by yzh119 — APPROVED** [review]
